[4.5/4.6/4.7/4.8 Regression] out of memory while parsing array with many initializers

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=14179

[4.5/4.6/4.7/4.8 regression] Uses lots of memory when compiling large initialized arrays

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=12245

--- trunk/gcc/ChangeLog
+++ trunk/gcc/ChangeLog
@@ -1,6 +1,18 @@
 	* config/rs6000/rs6000.md (call_value_indirect_aix32): Fix typo
 	in mode of operand[4].
 
+2012-01-16  Jan Hubicka  <jh@suse.cz>
+
+	PR c/12245
+	PR c++/14179
+	* convert.c (convert_to_integer): Use fold_convert for
+	converting an INTEGER_CST to integer type.
+
+2012-01-16  Jason Merrill  <jason@redhat.com>
+
+	PR c++/14179
+	* vec.c (vec_gc_o_reserve_1): Use ggc_round_alloc_size.
+
 2012-01-15  Uros Bizjak  <ubizjak@gmail.com>
 
 	PR rtl-optimization/51821
--- trunk/gcc/vec.c	2012/01/16 15:47:07	183212
+++ trunk/gcc/vec.c	2012/01/16 16:40:26	183213
@@ -221,6 +221,7 @@
 {
   struct vec_prefix *pfx = (struct vec_prefix *) vec;
   unsigned alloc = calculate_allocation (pfx, reserve, exact);
+  size_t size;
 
   if (!alloc)
     {
@@ -229,7 +230,17 @@
       return NULL;
     }
 
-  vec = ggc_realloc_stat (vec, vec_offset + alloc * elt_size PASS_MEM_STAT);
+  /* Calculate the amount of space we want.  */
+  size = vec_offset + alloc * elt_size;
+  /* Ask the allocator how much space it will really give us.  */
+  size = ggc_round_alloc_size (size);
+  /* Adjust the number of slots accordingly.  */
+  alloc = (size - vec_offset) / elt_size;
+  /* And finally, recalculate the amount of space we ask for.  */
+  size = vec_offset + alloc * elt_size;
+
+  vec = ggc_realloc_stat (vec, size PASS_MEM_STAT);
+
   ((struct vec_prefix *)vec)->alloc = alloc;
   if (!pfx)
     ((struct vec_prefix *)vec)->num = 0;
--- trunk/gcc/convert.c	2012/01/16 16:40:26	183213
+++ trunk/gcc/convert.c	2012/01/16 16:40:38	183214
@@ -860,6 +860,10 @@
 	  break;
 	}
 
+      /* When parsing long initializers, we might end up with a lot of casts.
+	 Shortcut this.  */
+      if (TREE_CODE (expr) == INTEGER_CST)
+	return fold_convert (type, expr);
       return build1 (CONVERT_EXPR, type, expr);
 
     case REAL_TYPE:

